packages/common/src/ContractUtils.ts

```diff
@@ -79,3 +80,79 @@ export const replaceLibraryBindingReferenceInArtitifact = <T>(artifact: T, libra
   const artifactString = JSON.stringify(artifact);
   return JSON.parse(artifactString.replace(/\$.*\$/g, libraryName));
 };
+
+export type EventSearchOptions = {
+  fromBlock: number;
+  toBlock: number;
+  filter?: any; // Object, allows caller to filter events by indexed paramateres. e.g. {filter: {myNumber: [12,13]}}
+  // filters all events where "myNumber" is 12 or 13.
+};
+
+/**
+ * Fetches specified contract event data for all input web3 providers. Returns false if any of the events found with
+ * one provider are NOT matched exactly in ALL of the other providers' event arrays.
+ * @param web3s Web3 providers to check for target event.
+ * @param contractAbi Contract ABI to query target event on.
+ * @param contractAddress Contract address to query target event on.
+ * @param eventName Name of target event.
+ * @param eventSearchOptions Target event search options. See here for more details: https://web3js.readthedocs.io/en/v1.5.2/web3-eth-contract.html#getpastevents
+ * @returns Object containing success of event query, missing events not found in all providers, and all event data
+ */
+export async function getEventsForMultipleProviders(
+  web3s: Web3[],
+  contractAbi: any[],
+  contractAddress: string,
+  eventName: string,
+  eventSearchOptions: EventSearchOptions
+): Promise<{ missingEvents: EventData[]; events: EventData[] }> {
+  const allProviderEvents = await Promise.all(
+    web3s.map((_web3) => {
+      const _contract = new _web3.eth.Contract(contractAbi, contractAddress);
+      return _contract.getPastEvents(eventName, eventSearchOptions);
+    })
+  );
+
+  // Associate each event with a key uniquely identifying the event. We'll use this key in the next step to determine
+  // which events were returned by all providers.
+  type UniqueEventData = { [key: string]: EventData };
+  const uniqueEventsForProvider: UniqueEventData[] = [];
+  // [index of web3Provider => {eventKey => event}]
+  const uniqueEvents: { [uniqueEventKey: string]: { event: EventData; count: number } } = {};
+  // Union map of ALL unique events across all providers: [{eventKey => {eventData, count}}]. We'll use this map to
+  // keep track of how many times we see each event. For an event to be returned successfully from this method, it must
+  // seen exactly once for each web3 provider.
+
+  const _getUniqueEventKey = (event: EventData): string => {
+    return JSON.stringify({
+      transactionHash: event.transactionHash,
+      transactionIndex: event.transactionIndex,
+      logIndex: event.logIndex,
+      returnValues: event.returnValues,
+      address: event.address,
+    });
+  };
+  allProviderEvents.forEach((eventDataForProvider, i) => {
+    uniqueEventsForProvider[i] = {};
+    eventDataForProvider.forEach((event) => {
+      const uniqueEventKey = _getUniqueEventKey(event);
+      // Add event for this provider.
+      uniqueEventsForProvider[i][uniqueEventKey] = event;
```

Comment:
Is this still needed?

Reply:
Nope! Great catch. This function is so much cleaner now